File: app/src/main/java/br/com/douglas/fukuhara/lodjinha/view/ProductDetailActivity.java
package br.com.douglas.fukuhara.lodjinha.view;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.text.Html;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.google.gson.Gson;

import br.com.douglas.fukuhara.lodjinha.R;
import br.com.douglas.fukuhara.lodjinha.interfaces.ProductDetailContract;
import br.com.douglas.fukuhara.lodjinha.network.RetrofitImpl;
import br.com.douglas.fukuhara.lodjinha.network.vo.ProductDataVo;
import br.com.douglas.fukuhara.lodjinha.presenter.ProductDetailPresenter;

public class ProductDetailActivity extends AppCompatActivity
        implements ProductDetailContract.View {

    private static final String PRODUCT_BUNDLE_KEY = "product_bundle_key";

    private ProductDetailContract.Presenter mPresenter;
    private ImageView mIvProductImage;
    private TextView mTvProductTitle;
    private TextView mTvProductPrevPrice;
    private TextView mTvProductFinalPrice;
    private TextView mTvProductDesc;
    private FloatingActionButton mFabReserve;

    public static Intent newIntent(Context context, ProductDataVo productDataVo) {
        Intent intent = new Intent(context, ProductDetailActivity.class);
        intent.putExtra(PRODUCT_BUNDLE_KEY, new Gson().toJson(productDataVo));
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_product_detail);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);

        ProductDataVo productDataVo = null;
        Bundle bundle = getIntent().getExtras();
        if (bundle != null && bundle.containsKey(PRODUCT_BUNDLE_KEY)) {
            productDataVo = new Gson().fromJson(bundle.getString(PRODUCT_BUNDLE_KEY), ProductDataVo.class);
            getSupportActionBar().setTitle(productDataVo.getCategoria().getDescricao());
        }

        setLayoutViews();

        mPresenter = new ProductDetailPresenter(productDataVo,
                this, RetrofitImpl.getInstance());
        mPresenter.displayProductDetail();
    }

    private void setLayoutViews() {
        mIvProductImage = findViewById(R.id.iv_product_image);
        mTvProductTitle = findViewById(R.id.tv_product_title);
        mTvProductPrevPrice = findViewById(R.id.tv_product_prev_price);
        mTvProductFinalPrice = findViewById(R.id.tv_product_final_price);
        mTvProductDesc = findViewById(R.id.tv_product_description);
        mFabReserve = findViewById(R.id.fab_product_reserve);

        mFabReserve.setOnClickListener(v -> mPresenter.onReserveButtonPressed());
    }

    @Override
    public void populateProductLayout(String urlImagem, String nome, String precoDe, String precoPor, String descricao) {
        Glide
                .with(this)
                .load(urlImagem)
                .error(R.mipmap.image_placeholder_error)
                .into(mIvProductImage);
        mTvProductTitle.setText(nome);
        mTvProductPrevPrice.setText(getString(R.string.best_seller_prev_price, precoDe));
        mTvProductFinalPrice.setText(getString(R.string.best_seller_final_price, precoPor));

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            mTvProductDesc.setText(Html.fromHtml(descricao, Html.FROM_HTML_MODE_COMPACT));
        } else {
            mTvProductDesc.setText(Html.fromHtml(descricao));
        }
    }

    @Override
    public void displayNoContentAvailable() {

    }

    @Override
    public void setFabEnabled(boolean enabled) {
        mFabReserve.setEnabled(enabled);
    }

    @Override
    public void showReservationSuccess() {
        buildAlertDialog(getString(R.string.dialog_product_reservation_success));
    }

    @Override
    public void showGenericFailureOnReservation() {
        buildAlertDialog(getString(R.string.dialog_product_reservation_fail));
    }

    @Override
    public void showFailureOnReservation(String serverMessage) {
        buildAlertDialog(serverMessage);
    }

    private void buildAlertDialog(String stringMessage) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder
                .setMessage(stringMessage)
                .setPositiveButton(R.string.dialog_ok_button, (dialog, which) -> mPresenter.onDialogConfirmationPressed() )
                .setCancelable(false)
                .create()
                .show();
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    @Override
    protected void onDestroy() {
        mPresenter.disposeAll();
        super.onDestroy();
    }
}

File: app/src/main/java/br/com/amedigital/challenge_android/view/ui/product_detail/ProductDetailActivity.kt
package br.com.amedigital.challenge_android.view.ui.product_detail

import android.content.DialogInterface
import android.graphics.Paint
import android.os.Build
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import br.com.amedigital.challenge_android.R
import br.com.amedigital.challenge_android.models.entity.Produto
import com.squareup.picasso.Picasso
import dagger.android.AndroidInjection
import kotlinx.android.synthetic.main.activity_product_detail.*
import kotlinx.android.synthetic.main.app_bar_home.*
import org.jetbrains.anko.backgroundColor

class ProductDetailActivity  : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_product_detail)

        val produto = intent.getParcelableExtra("produto") as? Produto

        toolbarTitle.text = ""

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            toolbarTitle.backgroundColor = getColor(R.color.greyish)
            toolbar.backgroundColor = getColor(R.color.greyish)
            ll_navi.backgroundColor = getColor(R.color.greyish)
        }
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)



        Picasso.with(this).load(produto?.urlImagem).into(productImageDeail)
        productTitleDetail.text = produto?.descricao
        productValueFrom.text = "De: " + produto?.precoDe.toString()

        productValueFrom.paintFlags = Paint.STRIKE_THRU_TEXT_FLAG
        productValueBy.text = "Por: " + produto?.precoPor.toString()

        productNameDetail.text = produto?.nome
        productDescription.text = produto?.descricao
        fab.setOnClickListener {
            var alert = AlertDialog.Builder(this)
            alert.setTitle("Produto reservado com sucesso")
            alert.setPositiveButton("ok"){ dialog, which ->
            }
            alert.show()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

}